### PR TITLE
ccall: optimize symbol lookup with redirection

### PR DESCRIPTION
**Pull request by vtjnash**
A followup to #42151

Also fixes a minor issue with jl_RTLD_DEFAULT_handle on Windows introduced by 653fb16c320.

### CODE REVIEW

File: src/ccall.cpp [hunk 1]
```diff
@@ -1271,7 +1287,7 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
 #ifdef _OS_WINDOWS_
             if ((f_lib == JL_EXE_LIBNAME) || // preventing invalid pointer access
                 (f_lib == JL_LIBJULIA_INTERNAL_DL_LIBNAME) ||
-                (!strcmp(f_lib, JL_LIBJULIA_DL_LIBNAME)) ||
+                (f_lib == JL_LIBJULIA_DL_LIBNAME) ||
```
Comment (staticfloat) [inline src/ccall.cpp:1290]: Is this guaranteed?  What happens if someone does `ccall((foo, "libjulia"), ...)`?

File: src/runtime_ccall.cpp [hunk 2]
```diff
@@ -36,7 +36,7 @@ void *jl_get_library_(const char *f_lib, int throw_err)
         return jl_exe_handle;
     if (f_lib == JL_LIBJULIA_INTERNAL_DL_LIBNAME)
         return jl_libjulia_internal_handle;
-    if (strcmp(f_lib, JL_LIBJULIA_DL_LIBNAME) == 0)
+    if (f_lib == JL_LIBJULIA_DL_LIBNAME)
```
Comment (staticfloat) [inline src/runtime_ccall.cpp:39]: Same question here

**Review by staticfloat — APPROVED** [review]

File: src/ccall.cpp (hunk 1 above)
Comment (vtjnash) [reply src/ccall.cpp:1290]: results are undefined (and always have been)

Comment (vtjnash) [reply src/ccall.cpp:1290]: but generally is supposed to prohibit this from triggering the optimizations here

**Review by JeffBezanson — APPROVED** [review]

**Pull request merged**